### ISSUE TITLE
Implement a check for missing encoding specification in test suite

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#18473]
gh-18468 fixed a longstanding issue with a missing encoding specification when reading a text file. That was usually file, but under some circumstances may cause a problem. https://github.com/scipy/scipy/pull/18468#issuecomment-1548778593 suggested a test suite approach to enforce the use of encodings more generally, and this would be a useful robustness improvement. Copying the comment:

Since the PR title mentions enforcement rather than just "using" utf-8, I figured I would mention how I solved the same general problem in meson: https://github.com/mesonbuild/meson/pull/10050

Sadly due to the intentional inefficiency of CPython itself, I needed to add runtime-installed code to the mesonbuild module to actually perform filtering -- `$PYTHONWARNINGS` and `python -W` are... not useless, but very crippled.

Basically, the testsuite sets up two environment variables so that:
- python emits warnings whenever you don't specify an encoding at all
- meson installs a warning filter to make EncodingWarning a fatal error

The testsuite will then fail with an exception every time `open` handles are created without specifying an encoding -- even for cases that aren't as easily greppable as `open(`, such as the two meson cases I fixed which used `ConfigParser().read()`.

**Comment by ilayn** [comment]
Convenience link : [PEP 686](https://peps.python.org/pep-0686/) which will fix this in the future
